lib/sps/detail/Sps.cpp

```diff
@@ -0,0 +1,235 @@
+		 _dm_low(dedispersion_plan.get_dm_low()),
+		 _dm_high(dedispersion_plan.get_dm_high()),
+		 _dm_step(dedispersion_plan.get_dm_step()),
+		 _in_bin(user_input.get_in_bin()),
+		 _out_bin(user_input.get_out_bin()),
+		 _input_buffer(input_data.get_input_buffer())
+{
+	_gpu_input_size 	= 0;
+	_d_input 			= NULL;
+	_gpu_output_size 	= 0;
+	_d_output 			= NULL;
+}
+
+template<typename SpsParameterType>
+Sps<SpsParameterType>::~Sps()
+{
+	cudaFree(_d_input);
+	cudaFree(_d_output);
+}
+
+
+template<typename SpsParameterType>
+void Sps<SpsParameterType>::allocate_memory_cpu_output(DedispersionPlan const &dedispersion_plan,
+													   float ****output_buffer,
+													   size_t *output_size)
+{
+	*output_buffer = (float ***) malloc(dedispersion_plan.get_range() * sizeof(float **));
+	for (int i = 0; i < dedispersion_plan.get_range(); ++i)
+	{
+		int total_samps = 0;
+		for (int k = 0; k < dedispersion_plan.get_num_tchunks(); ++k)
+			total_samps += dedispersion_plan.get_t_processed()[i][k];
+		(*output_buffer)[i] = (float **) malloc(dedispersion_plan.get_ndms()[i] * sizeof(float *));
+		for (int j = 0; j < dedispersion_plan.get_ndms()[i]; ++j)
+		{
+			(*output_buffer)[i][j] = (float *) malloc(( total_samps ) * sizeof(float));
+		}
+		*output_size += ( total_samps ) * dedispersion_plan.get_ndms()[i] * sizeof(float);
+	}
+}
+
+template<typename SpsParameterType>
+void Sps<SpsParameterType>::allocate_memory_gpu(DedispersionPlan const &dedispersion_plan)
+{
+	int time_samps = dedispersion_plan.get_t_processed()[0][0] + dedispersion_plan.get_maxshift();
+	_gpu_input_size = time_samps * (size_t) dedispersion_plan.get_nchans() * sizeof(unsigned short);
+
+	printf("\n gpu inputsize: %d\n", (int)(_gpu_input_size/1024/1024));
+	cudaError_t rc1 = ( cudaMalloc((void **)&_d_input, _gpu_input_size) );
+	if (rc1 != cudaSuccess)
+	    printf("Could not allocate memory: %d", rc1);
+
+	if (dedispersion_plan.get_nchans() < dedispersion_plan.get_max_ndms())
+	{
+		_gpu_output_size = time_samps * dedispersion_plan.get_max_ndms() * sizeof(float);
+	}
+	else
+	{
+		_gpu_output_size = time_samps * dedispersion_plan.get_nchans() * sizeof(float);
+	}
+	printf("\n gpu outputsize: %d\n", (int)(_gpu_output_size /1024/1024));
+	cudaError_t rc2 = ( cudaMalloc((void **)&_d_output, _gpu_output_size) );
+	if (rc2 != cudaSuccess)
+	    printf("Could not allocate memory: %d", rc2);
+
+
+	( cudaMemset(_d_output, 0, _gpu_output_size) );
+}
+
+template<typename SpsParameterType>
+void Sps<SpsParameterType>::operator()( unsigned device_id,
+										InputData &input_data,
+										DedispersionPlan &dedispersion_plan,
+										UserInput &user_input,
+										size_t gpu_memory,
+										std::vector<float> &output_sps,
+										float ****output_buffer,
+										size_t *output_size)
+{
+
+	//
+	long int inc = 0;
+	float tstart_local = 0.0f;
+
+	// allocate memory cpu output
+	allocate_memory_cpu_output(dedispersion_plan, output_buffer, output_size);
```

Comment:
how is this allocated memory freed - leak?
Instead of a float**** use a type.
just copy from: 
https://gitlab.com/SKA-TDT/cheetah/blob/master/cheetah/sps/cuda/DmTime.h
and memory will be released back on destruction of the object.